compiler/grainformat/grainformat.re

```diff
@@ -5,26 +5,61 @@ open Grain_parsing;
 open Grain_utils;
 open Filename;
 
+let file_eol: ref(option(Format.eol)) = ref(None);
+
+let determine_eol = line => {
+  let line_len = String.length(line);
+  if (line_len > 0) {
+    // check what the last char was
+    let last_char = line.[line_len - 1];
+    if (last_char == '\r') {
+      file_eol := Some(CRLF);
+    } else {
+      file_eol := Some(LF);
+    };
+  } else {
+    // must use OS default as this file has no newline we can use
+    file_eol :=
+      (
+        if (Sys.os_type == "Win32") {
+          Some(CRLF);
+        } else {
+          Some(LF);
+        }
+      );
+  };
+};
+
 let compile_parsed = (filename: option(string)) => {
   let program_str = ref("");
   let linesList = ref([]);
+  file_eol := None;
 
   switch (
     switch (filename) {
     | None =>
-      // force from stdin for now
-
       /* read from stdin until we get end of buffer */
       try(
         while (true) {
           let line = read_line();
+          switch (file_eol^) {
+          | None => determine_eol(line)
+          | Some(eol) => () // already set
+          };
           linesList := linesList^ @ [line];
         }
       ) {
       | exn => ()
       };
 
       program_str := String.concat("\n", linesList^);
+      // If this is a CRLF file add in the final \n after the \r that was removed by
+      // input_line
+
+      switch (file_eol^) {
+      | Some(CRLF) => program_str := program_str^ ++ "\n"
```

Comment:
It's because input_line removes the LFs so we end up with invalid code with the CR and Grain throws an error. We add it back and get a valid CRLF the lexer can handle.

Reply:
Just read the ocaml source code and it is really messy with this. Haha. Why do we read each line instead of the entire file since you are just concat'ing it anyway?

Reply:
Now reads the complete file if a file (but still needs to split it into constituent lines) or reads line by line over stdin

Reply:
I just checked piping files into stdin and we don't need to read line-by-line, so we can consolidate that logic

Reply:
Scratch that, I was incorrect. But I think we should just read the bytes instead of lines.